### DISCUSSION TITLE
Installing MPF on linux the right way

### DISCUSSION THREAD

**Discussion by gbouthenot** [missionpinball/missionpinball#115] [Show and tell]
This installs mpf, mpf-mc, mpf-language-server, mpf-monitor, vscode, vscode extension.

This is my cookbook to properly install MPF on debian-based linux (Debian, Ubuntu, Mint).
- Reduced sudo usage
- Custom python version
- NO python package systemwide
- Ability to switch between version really easily !

Benefits:
- If something go wrong, juste erase your virtualenv and start again.
- reusage virtualenv configuration for every team member, with exactly same versions

Tested on Debian 12, Linux Mint 19 & 21 on x86_64 architechiture. It does not work ARM (raspberry Pi)

# PYENV INSTALL
pyenv includes venv and deals with multiple python version. MPF currently works best with python 3.9, but recent debian ship with python 3.11

```
# https://github.com/pyenv/pyenv  / https://github.com/pyenv/pyenv-installer
wget https://github.com/pyenv/pyenv-installer/raw/master/bin/pyenv-installer
bash pyenv-installer
rm pyenv-installer

Shell customisation. Here it is for debian / mint. For other distro, maybe another file should be used

cat <<'EOF'>>.bashrc
export PYENV_ROOT="$HOME/.pyenv"
command -v pyenv >/dev/null || export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init -)"
EOF
exec $SHELL

## Check pyenv is available
pyenv
```

# PYTHON COMPILATION
```
# Dependencies:
sudo apt-get install libssl-dev libncurses-dev libffi-dev libreadline-dev libbz2-dev libsqlite3-dev liblzma-dev tk-dev

# python3.9 compilation
pyenv install 3.9.7 # (with patchs else 3.9.18 is the last python version)
```

# MPF INSTALLATION
```
# Dependencies for mpf monitor
sudo apt-get install libjpeg-dev libxcb-cursor0 libxkbcommon-x11-0 libxcb-icccm4 libxcb-keysyms1 libxcb-shape0 xsel
```

## 0.56 ENVIRONMENT
This is the version I use, it the currently stable versions.

```
# creation venv "0.56"
pyenv virtualenv 3.9.7 mpf-056

# Activation
pyenv activate 3.9.7/envs/mpf-056

# Packages install
pip3 install --upgrade pip
pip3 install mpf[cli]==0.56.2 mpf-mc==0.56.1 mpf-monitor==0.56.1 mpf-language-server==0.55.0.dev1

# exit virtual env
pyenv deactivate
```

# GUI INSTALLATION
Install a lightweight GUI
Don't do that if you already have a X-server installed.

```
sudo apt-get install lxde-core firefox-esr
```

# IDE INSTALLATION
```
cd /tmp
wget --trust-server-names https://go.microsoft.com/fwlink/?LinkID=760868
sudo apt install ./code_*.deb
rm code_*.deb
```

# MPF EXTENSION MPF FOR VSCODE

```
# https://github.com/missionpinball/mpf-ls/tree/dev/vscode-client
cd ~/data
git clone https://github.com/missionpinball/mpf-ls.git
(no need to be in a venv)
code --install-extension mpf-ls/vscode-client/mpfls-0.0.2.vsix
```

** How to use: **
Before launching `code`, you should activate your virtualenv:
```
pyenv activate 3.9.7/envs/mpf-056
code .
```

And color / syntax are working.

# Checking that everything work

```
# Copy mpf-examples
mkdir -p ~/data && cd ~/data
git clone https://github.com/missionpinball/mpf-examples.git
cd mpf-examples/
git branch -rv
git checkout origin/0.56.x -b 0.56.x
```

## modifying demos:
Before testing `demo-man`:
- Re-enable audio   it is currently disabled in the repository)
  - edit config/hardware.yaml (towards the end of file)
  ```
  sound_system
    enabled: true
  ```
- Re-enable BCP:
  - edit `config.yaml`, at the top of the file, comment:
  ```
  bcp:
    # connections: None
  ```

## Now we can test !

```
# activate our venv 0.56:
pyenv activate 3.9.7/envs/mpf-056

# Launch demos:
cd ~/data/mpf-examples/demo_man

mpf both -X
```

Media window should open. Check sound is OK (type `S` on the keyboard, then `L`, then `X`), you should see the Text UI.

Also try mpf-monitor:
in another shell:
```
pyenv activate 3.9.7/envs/mpf-056
cd ~/data/mpf-examples/demo_man
mpf monitor
```

Testing mc-demo:
To test the other example "mc_demo":
```
pyenv activate 3.9.7/envs/mpf-056
cd ~/data/mpf-examples/mc_demo
mpf both -X 
```

**Comment by gbouthenot** [comment]
ADDENTUM for P-Roc

# Install P-Rroc driver the right way

The only sudo make install, installs in /usr/local.
The only system modifications are /etc/dev/rules.

The python packages are not system-installed, and are installed by user, not root.

Download and unpack "debian installer 0.55":
```
wget https://github.com/missionpinball/mpf-debian-installer/archive/0.55.x.zip
unzip 0.55.x.zip
```

Install:
```
sudo apt-get install libudev-dev libusb-dev cmake

tar xf libusb-1.0.18.tar.bz2
tar xf libftdi1-1.1.tar.bz2
tar xf libusbx-1.0.18.tar.bz2
git clone --branch dev  https://github.com/preble/libpinproc libpinproc
# (the mpf repo is older: git clone --branch dev https://github.com/missionpinball/libpinproc.git libpinproc)
git clone https://github.com/missionpinball/pypinproc pypinproc

cd libusb-1.0.18
./configure
make
sudo make install
cd ..

cd libftdi1-1.1
mkdir -p bin && cd bin
cmake ..
make
sudo make install
cd ../..

cd libpinproc
mkdir -p bin && cd bin
cmake -DBUILD_SHARED_LIBS=ON ..
make
sudo make install
cd ../..

cd pypinproc
pyenv activate 3.9.7/envs/mpf-056
python3 setup.py install
cd ..

echo "export LD_LIBRARY_PATH=/usr/local/lib:\$LD_LIBRARY_PATH" >> ~/.bashrc
exec $SHELL

sudo cp 50-P-ROC.rules 51-P3-ROC.rules /etc/udev/rules.d/
sudo reboot
```

**Comment by elabree** [comment]
Thanks gbouthenot,
I struggled a lot installing Debian on VirtualBox. After failure of the Debian installer I tried to do a manual install. 
I got a runtime exception when starting demo_man (AttributeError: 'str'object has no attribute ítems'(bcp.py line 65).
So after a week I read your post and tried it. Now MPF runs fine with demo_man !
Awsome :-)

**Comment by gbouthenot** [comment]
Thank you for your kind message. I am glad this post has been useful !

**Comment by samuel49100** [comment]
Hi there, thanks **gbouthenot** for this complete installation tutorial !

Just a share if anyone has the same trouble, last week I tried to install a P-ROC driver board on Ubuntu 22.04 and got some "undefined symbol: **ftdi_read_data_set_chunksize**" from **pypinproc** module wrapping import. It's been solved by compiling **libftdi** from sources as you mention in your cookbook :)

**Comment by gbouthenot** [comment]
This is an updated method. I no longer compile and install libusb libftdi1-1.1.
This cause some incompatibilities. Please use the debian libs instead of these two.

This is the new method, it's better :

# python venv creation
```bash
env PYTHON_CONFIGURE_OPTS='--enable-optimizations --with-lto' PYTHON_CFLAGS='-march=native -mtune=native' pyenv install --verbose 3.11.5

pyenv virtualenv 3.11.5 mpf-057
pyenv activate mpf-057
pip3 install --upgrade pip

pip3 install mpf==0.57.3 mpf-mc==0.57.1 mpf-monitor==0.57.1
```

# Install P-ROC module:

```bash
sudo apt-get install cmake build-essential pkg-config libftdi1-2 libftdi1-dev

# les repos mpf sont plus récents que les repos "preble" !
cd ~/data
git clone https://github.com/missionpinball/libpinproc
cd libpinproc
mkdir -p bin && cd bin
cmake -DBUILD_SHARED_LIBS=ON ..
for f in CMakeFiles/pinproc.dir/link.txt CMakeFiles/pinprocfw.dir/link.txt; do sed -i -r -e "s/ -lusb / -lusb-1.0 /" $f ; done
make
sudo make install
cd ../..

git clone https://github.com/missionpinball/pypinproc
cd pypinproc
sed -i -r -e "s/'libusb'/'libusb-1.0'/" setup.py

# enable your venv (I use pyenv personally)
pyenv activate mpf-057

python3 setup.py install
cd ..

echo 'SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001", MODE:="0666", GROUP="plugdev"' | sudo tee /etc/udev/rules.d/50-P-ROC.rules
echo 'SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6015", MODE:="0666", GROUP="plugdev"' | sudo tee /etc/udev/rules.d/51-P3-ROC.rules
```

**Comment by gbouthenot** [comment]
Thank you for your message, I just posted an updated way to install it,
cleaner.
It's for Debian 12, though.
I don't recommend compiling usb & ftdi libraries.

merci !

Le lun. 16 déc. 2024 à 20:37, samuel49100 ***@***.***> a
écrit :

**Comment by AndyBadger** [comment]
Hello - I am having issues getting mpf-monitor to install on a raspberry pi.  Does anyone know how to get that working?  It seems to have issues with PyQt6.  I have mpf working and mpf-mc working.  But when i try to install mpf-monitor i cant get the install to work.

Also, note i have a Raspberry Pi5 and we will be using the FAST Neuron board system.

Thanks,
Andy

Here is what i see when i try the install:

(3.9.18/envs/mpf) pinball@raspberrypi:~/mpf-git/mpf-examples/demo_man $ pip install mpf-monitor
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting mpf-monitor
  Using cached https://www.piwheels.org/simple/mpf-monitor/mpf_monitor-0.57.1-py3-none-any.whl (37 kB)
Collecting PyQt6>=6.4.2 (from mpf-monitor)
  Using cached pyqt6-6.9.0.tar.gz (1.1 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [22 lines of output]
      Traceback (most recent call last):
        File "/home/pinball/.pyenv/versions/3.9.18/envs/mpf/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 389, in <module>
          main()
        File "/home/pinball/.pyenv/versions/3.9.18/envs/mpf/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 373, in main
          json_out["return_val"] = hook(**hook_input["kwargs"])
        File "/home/pinball/.pyenv/versions/3.9.18/envs/mpf/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 178, in prepare_metadata_for_build_wheel
          whl_basename = backend.build_wheel(metadata_directory, config_settings)
        File "/tmp/pip-build-env-bje0sqnq/overlay/lib/python3.9/site-packages/sipbuild/api.py", line 28, in build_wheel
          project = AbstractProject.bootstrap('wheel',
        File "/tmp/pip-build-env-bje0sqnq/overlay/lib/python3.9/site-packages/sipbuild/abstract_project.py", line 74, in bootstrap
          project.setup(pyproject, tool, tool_description)
        File "/tmp/pip-build-env-bje0sqnq/overlay/lib/python3.9/site-packages/sipbuild/project.py", line 629, in setup
          self.apply_user_defaults(tool)
        File "project.py", line 66, in apply_user_defaults
          super().apply_user_defaults(tool)
        File "/tmp/pip-build-env-bje0sqnq/overlay/lib/python3.9/site-packages/pyqtbuild/project.py", line 51, in apply_user_defaults
          super().apply_user_defaults(tool)
        File "/tmp/pip-build-env-bje0sqnq/overlay/lib/python3.9/site-packages/sipbuild/project.py", line 243, in apply_user_defaults
          self.builder.apply_user_defaults(tool)
        File "/tmp/pip-build-env-bje0sqnq/overlay/lib/python3.9/site-packages/pyqtbuild/builder.py", line 49, in apply_user_defaults
          raise PyProjectOptionException('qmake',
      sipbuild.pyproject.PyProjectOptionException
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
(3.9.18/envs/mpf) pinball@raspberrypi:~/mpf-git/mpf-examples/demo_man $
